feat: implement (optional) semver parsing of version

radeksimko opened the pull request:
Fixes #157 

I admit my JavaScript-fu is very rusty. I welcome any feedback. Also there's perhaps more changes outside of the context of the original problem - I'd be happy to extract these into separate PRs, if necessary.

crazy-max commented:
@radeksimko 

> I admit my JavaScript-fu is very rusty. I welcome any feedback.

Yeah me too hehe. I'll do a review soon. Thanks for your contribution!

crazy-max commented on package.json:
```diff
@@ -25,7 +25,9 @@
-    "@actions/tool-cache": "^1.5.5"
+    "@actions/tool-cache": "^1.5.5",
+    "@octokit/action": "^2.8.0",
+    "tmp": "^0.2.1"
```
`tmp` should be moved to `depDependencies`

crazy-max commented on src/github.ts:
```diff
@@ -1,12 +1,38 @@
+  if (!allTags) {
+    throw new Error(`Cannot find GoReleaser tags`);
+  }
+  console.log(`Found ${allTags.length} tags in total`);
```
Replace `console.log` with `core.debug`

crazy-max commented on src/github.ts:
```diff
@@ -1,12 +1,38 @@
 };
+
+const resolveVersion = async (version: string): Promise<string | null> => {
+  if (!process.env.GITHUB_TOKEN) {
```
I suppose this condition is in place to lift the restriction on the pagination of tags? I'm not a big fan of forcing the use of the GitHub token. I think it might be a good idea to have the list of tags available in JSON format on the goreleaser website. What do you think @caarlos0?

caarlos0 replied:
SGTM, we can probably generate it at build time 🤔

crazy-max replied:
@caarlos0 Yeah we can do that easily with [`actions/github-script`](https://github.com/actions/github-script) to extract all releases and create a PR with something like that:

```yaml
    - name: Fetch releases
      uses: actions/github-script@v2
      with:
        github-token: ${{secrets.GITHUB_TOKEN}}
        script: |
          const fs = require("fs");
          const response = await github.paginate(
            github.repos.listReleases.endpoint.merge({
              ...context.repo
            })
          );
          fs.writeFileSync(`www/docs/static/releases.json`, JSON.stringify(response));
    -
      name: Create Pull Request
      uses: peter-evans/create-pull-request@v2
      with:
        token: ${{ secrets.GITHUB_TOKEN }}
        commit-message: "chore(ci): update releases metadata"
        title: "chore(ci): update releases metadata"
        body: |
          Releases metadata updated.
        branch: releases-metadata
        branch-suffix: timestamp
        labels: automerge
```

caarlos0 replied:
DONE: https://goreleaser.com/static/releases.json

radeksimko commented:
@crazy-max I believe I addressed all your comments.

The implementation doesn't require GitHub token nor GitHub SDK anymore as it leverages the static JSON as discussed.

PTAL.

crazy-max commented:
@radeksimko LGTM!
